Change `start_time` and `end_time` handling in `combine_metadata`

pnuu opened the pull request:
The times of datasets used in composites were averaged to get the final time values for the composite. With this PR, the `start_time` and `end_time` attributes are instead changed to use the earliest and latest values, respectively. In addition, for `StaticImageCompositor` the default `start_time` and `end_time` values are set to `None` if they are not available in the filename.

 - [x] Closes #2630
 - [x] Closes #2734 
 - [x] Closes #2447
 - [x] Closes #2427 
 - [x] Closes #1174
 - [x] Closes #2446
 - [x] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

pnuu commented:
Going through the failing tests. The others are easy to fix, but I'm not sure what `combine_times=False` behaviour should be in `satpy.multiscene.stack()`. Looks like then the average is used, but would it ok to just skip that option and always use min/max of start/end times?

pnuu commented:
The `satpy.multiscene.stack()` option `combine_times` is not documented, so that would indicate that the deletion of it would be safe.

djhoese commented on satpy/composites/__init__.py:
```diff
@@ -1666,12 +1666,10 @@ def __call__(self, *args, **kwargs):
-        if "end_time" not in img.attrs or not img.attrs["end_time"]:
-            import datetime as dt
-            img.attrs["end_time"] = dt.datetime.utcnow()
+        if "start_time" not in img.attrs:
+            img.attrs["start_time"] = None
+        if "end_time" not in img.attrs:
+            img.attrs["end_time"] = None
```
Can you remind me, what does the `generic_image` reader do if it can't find start/end time from the filename? None or current time?

I guess if `None` then this isn't needed, right?

pnuu replied:
Right you are, seems to be `None`. I'll adjust when back on the work computer.

pnuu replied:
The `StaticImageCompositor` tests mock the image reading, so the times are not set. I'll remove these assertions.

djhoese commented on satpy/dataset/metadata.py:
```diff
@@ -27,20 +27,19 @@
-    is to handle cases where data were observed at almost the same time
-    but not exactly.  In the interest of time, lazy arrays are compared by
-    object identity rather than by their contents.
+    the returned dictionary.  The 'start_time' values will be set to the
+    earliest value and 'end_time' to latest time.  All other keys containing
+    the word 'time' are averaged.  Before these adjustments, non-datetime
+    objects are filtered out.  In the interest of time, lazy arrays are compared
+    by object identity rather than by their contents.
```
Should the `time_parameters` also get this min/max treatment:

https://satpy.readthedocs.io/en/stable/reading.html#time-metadata

pnuu replied:
I need to look at this in more detail later when I can concentrate a bit. I have a vague feeling that currently these are passed as-is.

djhoese replied:
Should this docstring be updated to reflect the `time_parameters`? Or are `X_start_time` and `X_end_time` not treated with min/max?

pnuu replied:
Yes. Updated in https://github.com/pytroll/satpy/pull/2737/commits/91dbc8e336af1c1830b6502f7accf488f1c90fd8

djhoese reviewed (approved):
Awesome job @pnuu! This looks good to me. It is obviously backwards incompatible, but I think this is the right path forward. I had a couple inline questions. The biggest one is what to do with `time_parameters` (see comment). Otherwise, I think the MultiScene removal of the combine_times makes sense. The only reason it existed was because it wasn't done in combine_metadata so if it happens in combine_metadata then great. It was only done in MultiScene because I wasn't sure if we wanted that behavior everywhere.

Some other concerns: What happens in `Scene.save_datasets` if start_time is None? I believe the default filename pattern includes `{start_time:%Y%m%d_%H%M%S}`. How do we want to handle that? Let it fail?

pnuu commented:
> What happens in Scene.save_datasets if start_time is None?
> How do we want to handle that? Let it fail?

The only case this should happen if the user is saving the plain data from `generic_image` reader that didn't have the time available in the filename. I guess we could add some kind of error handling for this case, but I'm not sure it's worth the effort :thinking:

djhoese commented:
> The only case this should happen if the user is saving the plain data from generic_image reader that didn't have the time available in the filename. I guess we could add some kind of error handling for this case, but I'm not sure it's worth the effort 🤔

Good point. I guess my only other fear would be odd situations with the MultiScene where you need to resample and have a static image, but the MultiScene wants to do something with ordering by start time...nah this shouldn't be a problem. Ok sounds good to not worry about it.

djhoese commented on satpy/dataset/metadata.py:
```diff
@@ -27,20 +27,23 @@
-    is to handle cases where data were observed at almost the same time
-    but not exactly.  In the interest of time, lazy arrays are compared by
-    object identity rather than by their contents.
+    the returned dictionary.  All values of the keys containing the substring
+    'start_time' will be set to the earliest value and similarly for 'end_time'
+    to latest time.  All other keys containing the word 'time' are averaged.
+    Before these adjustments, non-datetime objects are filtered out.
+
+    The same rules are applied to 'time_parameters' dictionary.
+
+    In the interest of processing time, lazy arrays are compared by object
+    identity rather than by their contents.
```
What about reorganizing so all the time stuff is in the same paragraph:

```suggestion
    the returned dictionary.
    
    All values of the keys containing the substring
    'start_time' will be set to the earliest value and similarly for 'end_time'
    to latest time.  All other keys containing the word 'time' are averaged.
    Before these adjustments, non-datetime objects are filtered out.
    The same rules are applied to the 'time_parameters' sub-dictionary.

    In the interest of processing time, lazy arrays are compared by object
    identity rather than by their contents.
```

Otherwise, I wanted to ask what does it mean "non-datetime objects are filtered out"? Does that mean keys with "time" in them that are not datetime objects are removed and not averaged? Are they removed even if they're the same?

pnuu replied:
Damn, didn't notice you had a suggestion here, already pushed my version. I also clarified the bit about "non-datetime objects"; it was meant mainly to say the `None` values coming from un-dated datasets were removed. And if there were any strings or such, they are also removed.

djhoese commented on satpy/dataset/metadata.py:
```diff
@@ -27,20 +27,25 @@
+    to the earliest value and similarly for 'end_time' to latest time.  All
+    other keys containing the word 'time' are averaged.  Before these adjustments,
+    `None` values resulting from data that don't have times associated to them
+    are removed. These rules are applied also to values in 'time_parameters'
```
```suggestion
    are removed. These rules are applied also to values in the 'time_parameters'
```

djhoese reviewed (approved):
I think this looks good. I think we should get @mraspaud, @sfinkens, @gerritholl, and @ameraner or @strandgren's opinions since this will affect all granule and segment based readers. I assume this group of developers have the widest experience with potential time-based edge cases.

djhoese commented:
Oh @pnuu I think this min/max code (including the time_parameters method it calls) could be removed:

https://github.com/pytroll/satpy/blob/cd33a0a5e17798a43e59efa4139f55448f9b4801/satpy/readers/file_handlers.py#L115-L118

sfinkens reviewed (approved):
Nice work, thanks!

pnuu commented on satpy/dataset/metadata.py:
```diff
@@ -75,17 +80,52 @@ def _shared_keys(info_dicts):
+        times = _combine_times(key, values)
+        if times is None:
+            return
+        shared_info[key] = times
```
Oh, indeed. Thanks! Fixed in https://github.com/pytroll/satpy/pull/2737/commits/3e050b0940fac573316f67a327010a97cba007da

pnuu commented:
> Oh @pnuu I think this min/max code (including the time_parameters method it calls) could be removed:
> 
> https://github.com/pytroll/satpy/blob/cd33a0a5e17798a43e59efa4139f55448f9b4801/satpy/readers/file_handlers.py#L115-L118

Removed the duplicate handling of times and adjusted the file handler test to actually use datetimes.

ameraner reviewed (approved):
LGTM, thanks for sorting this out! 
I don't see any problem regarding the segmented readers with this, since it seems to me that the previous behaviour in terms of min/max calculations for `start`, `nominal` and `observation` times is preserved. The segment sorting etc. is anyway not impacted by this, as it's based on chunk numbering from the single filehandlers.

Note: As discussed above, what still worries me a little bit is indeed the `generic_image` reader possibly returning datasets without a valid `start_time`... I think there are users that use satpy for simple operations like opening a geotiff, resample it and save it again, which could fail. Or maybe also applications like SIFT that may rely on a dataset having a `start_time`. But this goes outside the scope of this PR, which at least correctly fixed the composites misbehaviours.

gerritholl commented on satpy/dataset/metadata.py:
```diff
@@ -27,33 +27,37 @@
 from satpy.writers.utils import flatten_dict
 
 
-def combine_metadata(*metadata_objects, average_times=True):
```
Do we need a deprecation path, where a warning is raised if code passes `average_times`?

djhoese replied:
I'm not sure any people are actually using this, but given that it exists means we thought someone might want to control it so I agree that it should be documented at the very least. A specific deprecation warning would be nice to have.

The changes in the multiscene code are also backwards incompatible, but very very unlikely to be used by anyone except maybe Adam and Ernst. If I remember correctly the default behavior is preserved and was changed when the related kwarg was added to the multiscene stacking function. So my vote is no deprecation warning on the multiscene stuff, but warning on the metadata.py `average_times` would be nice to have.

pnuu replied:
I'll see about the deprecation warning, hopefully tomorrow.

pnuu replied:
I added a `UserWarning` if someone tries to use the `average_times` kwarg.

gerritholl commented on satpy/dataset/metadata.py:
```diff
@@ -27,33 +27,37 @@
+    `None` values resulting from data that don't have times associated to them
+    are removed. These rules are applied also to values in the 'time_parameters'
+    dictionary.
+
```
```suggestion
.. versionchanged:: 0.47
   Before Satpy 0.47, all times, including `start_time` and `end_time`, were averaged.
```

gerritholl reviewed (changes requested):
Thanks for this work!  There is a small risk that users will notice this backwards-incompatibility, so I have made a suggestion on explicitly mentioning in the documentation that the behaviour has changed, and (optionally) on raising a `DeprecrationWarning` or similar if a user does still pass `combine_times`.

djhoese commented:
> Note: As discussed above, what still worries me a little bit is indeed the generic_image reader possibly returning datasets without a valid start_time... I think there are users that use satpy for simple operations like opening a geotiff, resample it and save it again, which could fail. 

@ameraner good point, but skimming the changes in this PR again, I don't think the generic_image reader's behavior has changed at all. It was already returning a start_time of None and it was up to the user to override that...right?

djhoese commented on satpy/dataset/metadata.py:
```diff
@@ -27,33 +27,38 @@
+    other keys containing the word 'time' are averaged.  Before these adjustments,
+    `None` values resulting from data that don't have times associated to them
+    are removed. These rules are applied also to values in the 'time_parameters'
+    dictionary.
+.. versionchanged:: 0.47
+   Before Satpy 0.47, all times, including `start_time` and `end_time`, were averaged.
+    In the interest of processing time, lazy arrays are compared by object
+    identity rather than by their contents.
 
```
Looks like newlines and indentation didn't translate from the github comment. I think this is as I would expected it. Everything in the same indentation for the main docstring level then the versionchanged description is 3 spaces further indented (to align with the "v" in "versionchanged". I *think* that works. I don't remember between markdown and rst if it prefers 3 or 4 spaces for these.


```suggestion
    dictionary.
    
    .. versionchanged:: 0.47

       Before Satpy 0.47, all times, including `start_time` and `end_time`, were averaged.

    In the interest of processing time, lazy arrays are compared by object
    identity rather than by their contents.

```

pnuu replied:
Thanks, I'll generate the docs locally tomorrow to see it looks correct.

pnuu replied:
It seems that anything from one extra space works. I'll keep it at three for visual alignment.

ameraner commented:
> @ameraner good point, but skimming the changes in this PR again, I don't think the generic_image reader's behavior has changed at all. It was already returning a start_time of None and it was up to the user to override that...right?

Yes, indeed. Changing that would be outside the scope of this PR, and I'm not sure what the best solution would be anyway (since giving a "dummy" start_time can mess up other calculations, as we see here).

gerritholl reviewed (approved):
Thanks!

pnuu commented:
> > @ameraner good point, but skimming the changes in this PR again, I don't think the generic_image reader's behavior has changed at all. It was already returning a start_time of None and it was up to the user to override that...right?
> 
> Yes, indeed. Changing that would be outside the scope of this PR, and I'm not sure what the best solution would be anyway (since giving a "dummy" start_time can mess up other calculations, as we see here).

This could be handled in the writer (in another PR) with a simple (pseudo-code)
```python
if self.start_time is None:
    self.start_time = dt.datetime.utcnow()
fname = self.compose_fname_from_stuff()
```

djhoese commented:
> This could be handled in the writer (in another PR) with a simple (pseudo-code)

Eh, too much magic. If the start_time being None is a problem then the user should have to work around it. For example, if the filename generation is the problem then they should save it with a different filename template string.

mraspaud commented:
Everybody seems happy about this one, merging
